pandas/core/indexes/multi.py

```diff
@@ -3058,26 +3060,44 @@ def _update_indexer(idxr, indexer=indexer):
                 # a collection of labels to include from this level (these
                 # are or'd)
                 indexers = None
+                if not need_sort:
+                    # Find out if the list_like label are sorted as the levels or not
+                    k_codes = np.array(
+                        [self.levels[i].get_loc(e) for e in k if e in self.levels[i]]
+                    )
+                    need_sort = not (k_codes[:-1] < k_codes[1:]).all()
                 for x in k:
                     try:
                         idxrs = _convert_to_indexer(
                             self._get_level_indexer(x, level=i, indexer=indexer)
                         )
-                        indexers = idxrs if indexers is None else indexers | idxrs
+                        if need_sort and indexer is not None:
+                            # We intersect with indexer to make idxrs
+                            # ordered as previously seen indexes
+                            idxrs = indexer.intersection(idxrs)
+
+                        indexers = (
+                            idxrs
+                            if indexers is None
+                            else indexers.union(idxrs, sort=False)
+                        )
                     except KeyError:
-
                         # ignore not founds
                         continue
 
                 if indexers is not None:
-                    indexer = _update_indexer(indexers, indexer=indexer)
+                    if need_sort:
+                        indexer = indexers
+                    else:
+                        indexer = _update_indexer(indexers, indexer=indexer)
                 else:
                     # no matches we are done
                     return Int64Index([])._ndarray_values
 
             elif com.is_null_slice(k):
                 # empty slice
-                indexer = _update_indexer(None, indexer=indexer)
+                # index is given to conserve the order of this level
+                indexer = _update_indexer(Int64Index(np.arange(n)), indexer=indexer)
```

Comment:
I think you can remove this statement entirely, it actually does't do anything (and neither does the original)
it just yields the indexer

Reply:
This part define the comportement when having something like
`df.loc[(slice(None), [2, 1]), :]` (see my comment over it.
Other tests actually expect this to give the following results that keep the first level sorted exactly as existing, and the order of the subsequent level is then ignored.
One of this is test `TestMultiIndexSlicers.test_per_axis_per_level_doc_examples` in tests/indexings.
Keeping this behavior will lead to this kind of result:
```
     Ohio     Colorado
    Green Red    Green
a 1     0   1        2
  2     3   4        5
b 1     6   7        8
  2     9  10       11
```

I think the expected comportment can be one of this two:
```
# Ignoring the None slice order entirely
     Ohio     Colorado
    Green Red    Green
a 2     3   4        5
b 2     9  10       11
a 1     0   1        2
b 1     6   7        8

# Keeping the None slice order by code (and not by absolute position)
     Ohio     Colorado
    Green Red    Green
a 2     3   4        5
  1     0   1        2
b 2     9  10       11
  1     6   7        8
```

The first proposition is simpler to implement (ignore None slice as proposed).

Can you make a judgment call @jreback ?